File: android/cross-promo/src/main/kotlin/kr/dev/hoony/crosspromo/CrossPromo.kt
package kr.dev.hoony.crosspromo

import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.net.Uri
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kr.dev.hoony.crosspromo.data.CacheManager
import kr.dev.hoony.crosspromo.data.FirestoreService
import kr.dev.hoony.crosspromo.data.FrequencyManager
import kr.dev.hoony.crosspromo.internal.AppSelector
import kr.dev.hoony.crosspromo.model.PromoApp
import kr.dev.hoony.crosspromo.model.PromoConfig
import kr.dev.hoony.crosspromo.ui.CrossPromoDialog

object CrossPromo {

    private var currentAppId: String = ""
    private var isInitialized = false

    private lateinit var cacheManager: CacheManager
    private lateinit var frequencyManager: FrequencyManager
    private lateinit var firestoreService: FirestoreService

    private var apps: List<PromoApp> = emptyList()
    private var config: PromoConfig = PromoConfig()
    private var isDataLoaded = false

    /**
     * Initialize the SDK. Call this in Application.onCreate() or MainActivity.onCreate().
     *
     * @param context Application context
     * @param currentAppId The package name of the host app (e.g., "kr.dev.hoony.voda")
     * @param firebaseProjectId The Firebase project ID for CrossPromo Firestore (default: "crosspromosdk")
     */
    private var debugMode = false

    fun initialize(
        context: Context,
        currentAppId: String,
        firebaseProjectId: String = "crosspromosdk",
        debug: Boolean = false,
    ) {
        this.currentAppId = currentAppId
        this.debugMode = debug
        this.cacheManager = CacheManager(context.applicationContext)
        this.frequencyManager = FrequencyManager(context.applicationContext)
        this.firestoreService = FirestoreService(firebaseProjectId)
        this.isInitialized = true

        frequencyManager.incrementAppOpenCount()
        frequencyManager.resetSession()

        CoroutineScope(Dispatchers.IO).launch {
            loadData()
        }
    }

    private suspend fun loadData() {
        // Try cache first
        if (cacheManager.isCacheValid()) {
            apps = cacheManager.loadApps()
            config = cacheManager.loadConfig() ?: PromoConfig()
            if (apps.isNotEmpty()) {
                apps = AppSelector.filterApps(apps, currentAppId)
                isDataLoaded = true
                return
            }
        }

        // Fetch from Firestore
        try {
            val allApps = firestoreService.fetchApps()
            config = firestoreService.fetchConfig()

            // Cache raw data before filtering
            cacheManager.saveApps(allApps)
            cacheManager.saveConfig(config)

            // Filter for current platform and app
            apps = AppSelector.filterApps(allApps, currentAppId)
            isDataLoaded = true
        } catch (e: Exception) {
            // Use stale cache as fallback
            apps = AppSelector.filterApps(cacheManager.loadApps(), currentAppId)
            config = cacheManager.loadConfig() ?: PromoConfig()
            isDataLoaded = apps.isNotEmpty()
        }
    }

    /**
     * Check if a promo can be shown right now.
     */
    fun canShow(): Boolean {
        if (!isInitialized || !isDataLoaded) return false
        if (debugMode) return apps.isNotEmpty()
        if (!config.enabled) return false
        if (apps.isEmpty()) return false
        return frequencyManager.canShow(config.minAppOpenCount)
    }

    /**
     * Select an app to promote. Returns null if no app is available.
     */
    fun selectApp(): PromoApp? {
        if (!canShow()) return null
        return AppSelector.selectWeightedRandom(apps)
    }

    /**
     * Record that the promo was dismissed for today.
     */
    fun dismissForToday() {
        frequencyManager.dismissForToday()
    }

    /**
     * Record that the promo was shown in this session.
     */
    fun markShown() {
        frequencyManager.markSessionShown()
    }

    /**
     * Get the popup title from config.
     */
    fun getPopupTitle(): String = config.popupTitle

    /**
     * Open the Play Store page for the given app.
     */
    fun openPlayStore(context: Context, appId: String) {
        try {
            context.startActivity(
                Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=$appId")).apply {
                    addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                }
            )
        } catch (e: ActivityNotFoundException) {
            context.startActivity(
                Intent(
                    Intent.ACTION_VIEW,
                    Uri.parse("https://play.google.com/store/apps/details?id=$appId")
                ).apply {
                    addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                }
            )
        }
    }

    /**
     * Composable helper: shows the cross-promo dialog if conditions are met.
     * Place this in your screen composable.
     *
     * Usage:
     * ```
     * CrossPromo.ShowIfNeeded()
     * ```
     */
    @Composable
    fun ShowIfNeeded(
        delayMillis: Long = 2000L,
    ) {
        if (!isInitialized) return

        var showDialog by remember { mutableStateOf(false) }
        var selectedApp by remember { mutableStateOf<PromoApp?>(null) }

        LaunchedEffect(Unit) {
            kotlinx.coroutines.delay(delayMillis)
            if (canShow()) {
                selectedApp = selectApp()
                if (selectedApp != null) {
                    showDialog = true
                }
            }
        }

        if (showDialog && selectedApp != null) {
            val context = androidx.compose.ui.platform.LocalContext.current
            CrossPromoDialog(
                app = selectedApp!!,
                title = getPopupTitle(),
                onInstall = {
                    openPlayStore(context, selectedApp!!.appId)
                    markShown()
                    showDialog = false
                },
                onDismissToday = {
                    dismissForToday()
                    showDialog = false
                },
                onClose = {
                    markShown()
                    showDialog = false
                },
            )
        }
    }
}

File: ios/Sources/CrossPromoSDK/CrossPromo.swift
import SwiftUI
import StoreKit

public final class CrossPromo: ObservableObject {

    public static let shared = CrossPromo()

    private var currentAppId: String = ""
    private var firestoreService: FirestoreService!
    private let cacheManager = CacheManager()
    private let frequencyManager = FrequencyManager()

    private var apps: [PromoApp] = []
    @Published private var config = PromoConfig()
    private var isDataLoaded = false

    private init() {}

    // MARK: - Public API

    /// Initialize the SDK. Call this in your App's init() or on appear.
    ///
    /// - Parameters:
    ///   - currentAppId: The bundle identifier of the host app
    ///   - firebaseProjectId: The Firebase project ID for CrossPromo Firestore (default: "crosspromosdk")
    public func initialize(currentAppId: String, firebaseProjectId: String = "crosspromosdk") {
        self.currentAppId = currentAppId
        self.firestoreService = FirestoreService(projectId: firebaseProjectId)
        frequencyManager.incrementAppOpenCount()
        frequencyManager.resetSession()

        Task {
            await loadData()
        }
    }

    /// Check if a promo can be shown right now.
    public func canShow() -> Bool {
        guard isDataLoaded, config.enabled, !apps.isEmpty else { return false }
        return frequencyManager.canShow(minAppOpenCount: config.minAppOpenCount)
    }

    /// Select an app to promote. Returns nil if no app is available.
    public func selectApp() -> PromoApp? {
        guard canShow() else { return nil }
        return AppSelector.selectWeightedRandom(apps)
    }

    /// Record that the promo was dismissed for today.
    public func dismissForToday() {
        frequencyManager.dismissForToday()
    }

    /// Record that the promo was shown in this session.
    public func markShown() {
        frequencyManager.markSessionShown()
    }

    /// Get the popup title from config.
    public var popupTitle: String { config.popupTitle }

    /// Open the App Store page for the given app.
    public func openAppStore(for app: PromoApp) {
        if let appStoreId = Int(app.iosAppStoreId), appStoreId > 0 {
            openStoreProduct(appStoreId: appStoreId)
        } else if let urlString = app.storeLinks["ios"],
                  let url = URL(string: urlString) {
            UIApplication.shared.open(url)
        }
    }

    // MARK: - Private

    private func loadData() async {
        if cacheManager.isCacheValid {
            let cachedApps = cacheManager.loadApps()
            if let cachedConfig = cacheManager.loadConfig(), !cachedApps.isEmpty {
                config = cachedConfig
                apps = AppSelector.filterApps(cachedApps, currentAppId: currentAppId)
                isDataLoaded = true
                return
            }
        }

        do {
            let allApps = try await firestoreService.fetchApps()
            let fetchedConfig = try await firestoreService.fetchConfig()

            cacheManager.saveApps(allApps)
            cacheManager.saveConfig(fetchedConfig)

            await MainActor.run {
                config = fetchedConfig
                apps = AppSelector.filterApps(allApps, currentAppId: currentAppId)
                isDataLoaded = true
            }
        } catch {
            let cachedApps = cacheManager.loadApps()
            let cachedConfig = cacheManager.loadConfig() ?? PromoConfig()
            await MainActor.run {
                config = cachedConfig
                apps = AppSelector.filterApps(cachedApps, currentAppId: currentAppId)
                isDataLoaded = !apps.isEmpty
            }
        }
    }

    private func openStoreProduct(appStoreId: Int) {
        guard let scene = UIApplication.shared.connectedScenes
            .compactMap({ $0 as? UIWindowScene })
            .first,
              let rootVC = scene.windows.first?.rootViewController else {
            return
        }

        let storeVC = SKStoreProductViewController()
        storeVC.loadProduct(withParameters: [
            SKStoreProductParameterITunesItemIdentifier: NSNumber(value: appStoreId)
        ])

        var topVC = rootVC
        while let presented = topVC.presentedViewController {
            topVC = presented
        }
        topVC.present(storeVC, animated: true)
    }
}

// MARK: - SwiftUI View Modifier

public extension View {

    /// Show the cross-promotion popup if conditions are met.
    ///
    /// Usage:
    /// ```swift
    /// ContentView()
    ///     .crossPromoPopup()
    /// ```
    func crossPromoPopup(delaySeconds: Double = 2.0) -> some View {
        modifier(CrossPromoModifier(delaySeconds: delaySeconds))
    }
}

struct CrossPromoModifier: ViewModifier {
    let delaySeconds: Double

    @State private var showPopup = false
    @State private var selectedApp: PromoApp?

    func body(content: Content) -> some View {
        content
            .onAppear {
                DispatchQueue.main.asyncAfter(deadline: .now() + delaySeconds) {
                    if CrossPromo.shared.canShow() {
                        selectedApp = CrossPromo.shared.selectApp()
                        if selectedApp != nil {
                            showPopup = true
                        }
                    }
                }
            }
            .overlay {
                if showPopup, let app = selectedApp {
                    CrossPromoView(
                        app: app,
                        title: CrossPromo.shared.popupTitle,
                        onInstall: {
                            CrossPromo.shared.openAppStore(for: app)
                            CrossPromo.shared.markShown()
                            showPopup = false
                        },
                        onDismissToday: {
                            CrossPromo.shared.dismissForToday()
                            showPopup = false
                        },
                        onClose: {
                            CrossPromo.shared.markShown()
                            showPopup = false
                        }
                    )
                    .transition(.opacity)
                    .animation(.easeInOut(duration: 0.3), value: showPopup)
                }
            }
    }
}
